src/renderer/grid_renderer.rs

```diff
@@ -162,42 +148,34 @@ impl GridRenderer {
         &mut self,
         canvas: &Canvas,
         text: &str,
-        grid_position: (u64, u64),
-        cell_width: u64,
+        grid_position: GridPos<i32>,
+        cell_width: i32,
         style: &Option<Arc<Style>>,
     ) -> bool {
         tracy_zone!("draw_foreground");
-        let (x, y) = grid_position * self.font_dimensions;
-        let width = cell_width * self.font_dimensions.width;
+        let pos = grid_position.cast() * self.grid_scale;
+        let size = GridSize::new(cell_width, 0).cast() * self.grid_scale;
+        let width = size.width;
 
         let style = style.as_ref().unwrap_or(&self.default_style);
         let mut drawn = false;
 
         // We don't want to clip text in the x position, only the y so we add a buffer of 1
         // character on either side of the region so that we clip vertically but not horizontally.
-        let (grid_x, grid_y) = grid_position;
-        let clip_position = (grid_x.saturating_sub(1), grid_y);
+        let clip_position = (grid_position.x.saturating_sub(1), grid_position.y).into();
         let region = self.compute_text_region(clip_position, cell_width + 2);
 
         if let Some(underline_style) = style.underline {
-            let line_position = self.shaper.underline_position();
-            let p1 = (
-                x as f32,
-                (y - line_position + self.font_dimensions.height) as f32,
-            );
-            let p2 = (
-                (x + width) as f32,
-                (y - line_position + self.font_dimensions.height) as f32,
-            );
+            let line_position = self.grid_scale.0.height - self.shaper.underline_position();
+            let p1 = pos + PixelVec::new(0.0, line_position);
+            let p2 = pos + PixelVec::new(width, line_position);
 
-            self.draw_underline(canvas, style, underline_style, p1.into(), p2.into());
+            self.draw_underline(canvas, style, underline_style, p1, p2);
             drawn = true;
         }
 
         canvas.save();
-        canvas.clip_rect(region, None, Some(false));
-
-        let y_adjustment = self.shaper.y_adjustment();
+        canvas.clip_rect(to_skia_rect(&region), None, Some(false));
```

Comment:
same here. Imho we should have to_skia be a method here rather than a free function